### DISCUSSION TITLE
Controlling styling / modifications to D3FC charts

### DISCUSSION THREAD

**Discussion by limx0** [perspective-dev/perspective#2200] [General]
I'm currently streaming data from the tornado backend to a couple of perspective widgets in a html file (thank you for the plentiful examples!) one of which is a rolling time series scatter plot with multiple series. Currently, I believe, the mapping of colours/markers in the `X/Y scatter` plot is basically just the enumeration of whatever data exists in the table at a given time. This is causing my series to change colour/marker if one of the columns in the table is all null for some period of time. Is it possible to declare these styling attributes via some other config?

I wonder if it is within scope for perspective to allow configuring the plots from table values (for example like holoviews does - https://holoviews.org/user_guide/Style_Mapping.html)? 

An example (the labels of L2 and L3 will flicker between orange and green depending on what is present in the data):

Set x-axis = `timestamp`, y-axis=`value`
```
import asyncio
import perspective
import random
import datetime

async def update(table):
    while True:
        now = datetime.datetime.now()
        data = [{"value": 1, "label": "L1", "timestamp": now}]
        for label in ("L1", "L2", "L3"):
            val = random.random()
            if val > 0.9:
                data.append({"value": val, "label": label, "timestamp": now})
        table.update(data)
        await asyncio.sleep(0.1)
        

schema = {
    "timestamp": datetime.datetime,
    'value': float,
    'label': str,
}

table = perspective.Table(schema, limit=10)
widget = perspective.PerspectiveWidget(table,  column_pivots=['label'], plugin="X/Y Scatter")
asyncio.get_event_loop().create_task(update(table))
widget # show
```

**Comment by limx0** [comment]
Hey @texodus - sorry for the tag but I see you've been hard at work on some UI improvements (which look awesome!) so I just wanted to bump and see if anything like this was on the radar (specifically more control over D3FC via table values ala holoviews Style Mapping link above)? 

I could potentially offer a bounty for this work if that would help this along?

**Comment by texodus** [comment]
@limx0 there three conflated issues present we'd like to address:
1) Color range of `"split_by"` (formerly `"column_pivots"`) shouldn't re-assign when the table is updated. This is a bug.
2) `"X/Y Scatter"` _also has_ a color field (column position 3), so you could use `"label"` column here, but it does not currently work with string columns. However, `"Treemap"` and `"Sunburst"` do work correctly with string columns in the color field, so this is another easy bug.
3) We're working on a few UX features to specify the styling properties of charts, but this work is still in the design stage.

I've opened issues #2198 (1) and #2199 (2) to track these bugs, which should at least fix your issue for now, so I'm going to convert this issue to a discussion, in favor of those for tracking.

**Reply by limx0** [reply]
Thanks for the update and the issues @texodus! I will have a look at the `X/Y Scatter` colour field also.
